### PR TITLE
dialects: dmp: make strategies attributes and carry them in IR

### PR DESCRIPTION
**Pull request by PapyChacal**
As of now, the decomposition strategies are dataclasses and the information to construct them is expected to match in multiple bits of the pipeline (actually fused in `distribute-stencil`, which happens to do shape inference too, but I'm trying to decouple things here)

So instead, define them as attributes, and make the initial distribution pass pop that attribute on the swaps. The swap shape inference then do not need extra information to carry on.

Teaser: I want this part to be just collaborating shape inference as in:
- #3047

### CODE REVIEW

File: tests/filecheck/transforms/distribute-stencil.mlir [hunk 1]
```diff
@@ -26,7 +26,7 @@ builtin.module {
...
 // CHECK-NEXT:      %3 = stencil.load %0 : !stencil.field<[-4,68]x[-4,68]x[-4,68]xf64> -> !stencil.temp<[-1,33]x[-1,33]x[0,32]xf64>
-// CHECK-NEXT:      "dmp.swap"(%3) {"topo" = #dmp.topo<2x2x2>, "swaps" = [#dmp.exchange<at [32, 0, 0] size [1, 32, 32] source offset [-1, 0, 0] to [1, 0, 0]>, #dmp.exchange<at [-1, 0, 0] size [1, 32, 32] source offset [1, 0, 0] to [-1, 0, 0]>, #dmp.exchange<at [0, 32, 0] size [32, 1, 32] source offset [0, -1, 0] to [0, 1, 0]>, #dmp.exchange<at [0, -1, 0] size [32, 1, 32] source offset [0, 1, 0] to [0, -1, 0]>]} : (!stencil.temp<[-1,33]x[-1,33]x[0,32]xf64>) -> ()
+// CHECK-NEXT:      "dmp.swap"(%3) {"topo" = #dmp.topo<2x2x2>, "strat" = #dmp.grid_slice_2d<[#builtin.int<2>, #builtin.int<2>, #builtin.int<2>], false>, "swaps" = [#dmp.exchange<at [32, 0, 0] size [1, 32, 32] source offset [-1, 0, 0] to [1, 0, 0]>, #dmp.exchange<at [-1, 0, 0] size [1, 32, 32] source offset [1, 0, 0] to [-1, 0, 0]>, #dmp.exchange<at [0, 32, 0] size [32, 1, 32] source offset [0, -1, 0] to [0, 1, 0]>, #dmp.exchange<at [0, -1, 0] size [32, 1, 32] source offset [0, 1, 0] to [0, -1, 0]>]} : (!stencil.temp<[-1,33]x[-1,33]x[0,32]xf64>) -> ()
```
Comment (n-io) [inline tests/filecheck/transforms/distribute-stencil.mlir:29]: Is there duplicate information between strat and topo?

Comment (PapyChacal) [reply tests/filecheck/transforms/distribute-stencil.mlir:29]: @AntonLydike Do you have an opinion on this?
My understanding is that they do not represent the same thing (rank topology and gride slicing respectively) but that those happen to coincide in the current strategies.
I'd be fine just making the topo part of the strategy or keep as-is: do you think it's worth keeping them decoupled?

Comment (PapyChacal) [reply tests/filecheck/transforms/distribute-stencil.mlir]: Okay, went with making topo a part of strategy anyway.
As pointed by @n-io, returning a topology is part of a strategy's interface anyway!

File: xdsl/dialects/experimental/dmp.py [hunk 2]
```diff
@@ -426,6 +427,160 @@ def print_parameters(self, printer: Printer) -> None:
...
+    """
+    Takes a grid with two or more dimensions, slices it along the first two into equally
+    sized segments.
+    """
+
+    name = "dmp.grid_slice_2d"
+
+    topology: ParameterDef[RankTopoAttr]
+
+    diagonals: ParameterDef[builtin.BoolAttr]
```
Comment (superlopuh) [inline xdsl/dialects/experimental/dmp.py:452]: Would this make sense as an optional unitattr? the IR would be a bit more self-explanatory than just `false`

File: xdsl/dialects/experimental/dmp.py [hunk 3]
```diff
@@ -426,6 +427,160 @@ def print_parameters(self, printer: Printer) -> None:
...
+            ),
+            *(size for size in shape[2:]),
+        )
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        yield from _flat_face_exchanges_for_dim(shape, 0)
+
+        yield from _flat_face_exchanges_for_dim(shape, 1)
+
+        # TOOD: add diagonals
```
Comment (superlopuh) [inline xdsl/dialects/experimental/dmp.py:483]: TOOD

Comment (superlopuh) [reply xdsl/dialects/experimental/dmp.py:483]: But also probably better to `raise NotImplementedError("Diagonals not yet implemented")`

File: xdsl/dialects/experimental/dmp.py [hunk 4]
```diff
@@ -426,6 +427,160 @@ def print_parameters(self, printer: Printer) -> None:
...
+        )
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        yield from _flat_face_exchanges_for_dim(shape, 0)
+
+        yield from _flat_face_exchanges_for_dim(shape, 1)
+
+        yield from _flat_face_exchanges_for_dim(shape, 2)
+
+        # TOOD: add diagonals
```
Comment (superlopuh) [inline xdsl/dialects/experimental/dmp.py:533]: TOOD!

Comment (PapyChacal) [reply xdsl/dialects/experimental/dmp.py:533]: I thought you ondly didn't like TODOs? :stuck_out_tongue_winking_eye: 

Just moved the code, will polish that

File: xdsl/dialects/experimental/dmp.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/dialects/experimental/dmp.py:452]: How?
https://xdsl.zulipchat.com/#narrow/stream/364199-Beginners/topic/Optional.20attribute.20parameter.20def

Comment (PapyChacal) [reply xdsl/dialects/experimental/dmp.py:452]: I'm also not trying to perfect the syntax just yet here!

Comment (superlopuh) [reply xdsl/dialects/experimental/dmp.py:452]: OK then please disregard :)

**Review by n-io — APPROVED** [review]

**Pull request merged**